ecs-agent/modeltransformer/transformer.go

```diff
@@ -0,0 +1,106 @@
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package modeltransformer
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger"
+)
+
+const (
+	modelTypeTask = "Task"
+)
+
+// Transformer stores transformation functions for all types of objects.
+// Transform<type> will execute a series of transformation functions to make it compatible with current agent version.
+// add<type>TransformationFunctions will add more <type> transformation functions to the transformation functions chain.
+// Add other transformation functions as needed. e.g. ContainerTransformationFunctions.
+// Add corresponding Transform<Type> and Add<Type>TransformationFunctions while adding other transformation functions.
+type Transformer struct {
+	taskTransformFunctions []*TransformFunc
+}
+
+type transformationFunctionClosure func([]byte) ([]byte, error)
+
+// TransformFunc contains the threshold version string for transformation function and the transformationFunction itself.
+// During upgrade, all models from versions below threshold version should execute the transformation function.
+type TransformFunc struct {
+	version  string
+	function transformationFunctionClosure
+}
+
+func NewTransformer() *Transformer {
+	t := &Transformer{}
+	return t
+}
+
+// GetNumberOfTransformationFunctions returns the number of transformation functions given a model type
+func (t *Transformer) GetNumberOfTransformationFunctions(modelType string) int {
+	switch modelType {
+	case modelTypeTask:
+		return len(t.taskTransformFunctions)
+	default:
+		return 0
+	}
+}
+
+// TransformTask executes the transformation functions when version associated with model in boltdb is below the threshold
+func (t *Transformer) TransformTask(version string, data []byte) ([]byte, error) {
+	var err error
+	// execute transformation functions sequentially and skip those not applicable
+	for _, transformFunc := range t.taskTransformFunctions {
+		if checkVersionSmaller(version, transformFunc.version) {
+			logger.Info(fmt.Sprintf("Agent version associated with task model in boltdb %s is below threshold %s. Transformation needed.", version, transformFunc.version))
+			data, err = transformFunc.function(data)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			logger.Info(fmt.Sprintf("Agent version associated with task model in boltdb %s is bigger or equal to threshold %s. Skipping transformation.", version, transformFunc.version))
+			continue
+		}
+	}
+	return data, err
+}
+
+// AddTaskTransformationFunctions adds the transformationFunction to the handling chain
+func (t *Transformer) AddTaskTransformationFunctions(version string, transformationFunc transformationFunctionClosure) {
```

Comment:
Q. Should we add input validation, such as validating Agent version is not an empty string, before add it to the function list?

Reply:
Similar as above, this is not a user input but the code we write, I would not think doing such validation is required? What's your thought?

Reply:
Updated.